### ISSUE TITLE
Input layer doesn't properly switch between data readers during LTFB

### ISSUE THREAD

**Issue by timmoon10** [LLNL/lbann#1597]
While debugging #1596, I encountered errors with the message:

```
Number of samples: 244 does not match the number of responses: 200
```

The error comes from:

https://github.com/LLNL/lbann/blob/95474f70114ccef94ca6194febf9e566bfaecf43/include/lbann/io/data_buffers/generic_io_buffer.hpp#L62

I suspect the error happens because LTFB switches between execution modes without reseting background fetching in the IO buffers. If I understand correctly, the input layer works like:

```c++
void fp_compute() {
  ++m_buffer_idx;
  m_io_buffers[m_buffer_idx % m_io_buffers.size()].get_data();
  m_io_buffers[(m_buffer_idx+1) % m_io_buffers.size()].fetch_data_in_background();
}
```

If we don't reset the IO buffers when switching execution modes, then the input layer receives data that was fetched asynchronously from the wrong data reader. If an LTFB round happens at step 10 and there are 11 steps/epoch, the 11th training step will expect a partial mini-batch from the training set but receive a full mini-batch from the validation set. This mismatch triggers our error.

Do we already have machinery to reset the IO buffers? During the LTFB round, it seems that the evaluation epochs are correctly using the validation set. It's only when we finish the evaluation and go back to the middle of the training set that we appear to load the wrong data.

I'm not sure if this will help #1596. I wouldn't expect us to get divergence from a few mini-batches with incorrect data, unless it was filled with garbage.
